lib/puppet/datatypes.rb

```diff
@@ -0,0 +1,190 @@
+# Data types in the Puppet Language can have implementations written in Ruby
+# and distributed in puppet modules. A data type can be declared together with
+# its implementation by creating a file in 'lib/puppet/functions/<modulename>'.
+# The name of the file must be the downcased name of the data type followed by
+# the extension '.rb'.
+#
+# A data type is created by calling {Puppet::DataTypes.create_type(<type name>)}
+# and passing it a block that defines the data type interface and implementation.
+#
+# Data types are namespaced inside the modules that contains them. The name of the
+# data type is prefixed with the name of the module. As with all type names, each
+# segment of the name must start with an uppercase letter.
+#
+# @example A simple data type
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String
+#       }
+#     PUPPET
+#   end
+#
+# The above example does not declare an implementation which makes it exactly similar
+# to adding the following contents in a file named 'user.pp' under the 'types' directory
+# of the module root.
+#
+#   type Auth::User = Object[
+#     attributes => {
+#       name => String,
+#       email => String
+#     }]
+#
+# Both declarations are valid and will be found by the module loader.
+#
+# Anatomy of a data type
+# ---
+#
+# Data types consist of an interface and an implementation. A data type can often be used
+# by only declaring the interface. Unless an implementation is found it will be automatically
+# generated.
```

Comment:
Perhaps better to say "A Data Type consists of an interface and an implementation. Unless a registered implementation is found, the type system will automatically generate one. An automatically generated implementation is all that is needed when the interface fully defines the behaviour (for example in the common case when the data type has no other behaviour than having attributes). When the automatically generated implementation is not sufficient, one must be implemented and registered. The implementation can either be done next to the interface definition, or map to an existing implementation class. An implementation class does not have to be special in other respects than that it must implemented the type's interface. This makes it possible to use existing Ruby data types as data types in the puppet language." 

(This also replaces the sentence below).
I wrote the above suggestion because "often" and "in some case" left too much open to interpretation IMO.